File: src/domain/recommendation.ts
export type DemoStatus =
  | "demo_only"
  | "released_with_demo"
  | "released"
  | "unknown";

export type ReleaseBucket = "new" | "classic";

export interface GameRecommendationFacts {
  appid: number;
  name: string;
  releaseDate?: string | null;
  positiveReviewPct?: number | null;
  totalReviews?: number | null;
  currentPlayers?: number | null;
  multiplayerModes: string[];
  demoStatus: DemoStatus;
  aiScore?: number | null;
}

export function scoreGame(
  facts: GameRecommendationFacts,
  today: Date = new Date(),
): number {
  if (typeof facts.aiScore === "number" && Number.isFinite(facts.aiScore)) {
    return roundOne(clamp(facts.aiScore, 0, 100));
  }

  const reviewQuality = lightweightReviewQuality(facts);
  const multiplayerFit = lightweightMultiplayerFit(facts.multiplayerModes);
  const freshness = freshnessScore(facts.releaseDate, today);
  const discoveryValue = lightweightDiscoveryValue(
    facts,
    reviewQuality,
    freshness,
  );
  const confidence = lightweightConfidence(facts);
  const uncertaintyPenalty = (1 - confidence) * 10;
  const preanalysisPenalty = 4;

  const lightweightQualityProxy =
    0.45 * reviewQuality +
    0.3 * multiplayerFit +
    0.15 * freshness +
    0.1 * discoveryValue;

  return roundOne(
    clamp(
      0.55 * lightweightQualityProxy +
        0.2 * multiplayerFit +
        0.15 * discoveryValue +
        0.1 * freshness -
        uncertaintyPenalty -
        preanalysisPenalty,
      0,
      100,
    ),
  );
}

export function bucketGame(
  facts: Pick<GameRecommendationFacts, "releaseDate">,
  today: Date = new Date(),
): ReleaseBucket {
  const days = daysSinceRelease(facts.releaseDate, today);
  return days !== null && days >= 0 && days <= 30 ? "new" : "classic";
}

function lightweightReviewQuality(facts: GameRecommendationFacts): number {
  const rawPositiveRate = clamp(facts.positiveReviewPct ?? 0, 0, 100) / 100;
  const totalReviews = facts.totalReviews ?? 0;
  const positive = totalReviews * rawPositiveRate;
  const bayesPositiveRate = (positive + 35) / (totalReviews + 35 + 15);
  const confidence = 1 - Math.exp(-totalReviews / 120);

  return clamp(
    100 * (bayesPositiveRate * 0.85 + rawPositiveRate * 0.15) * confidence +
      55 * (1 - confidence),
    0,
    100,
  );
}

function lightweightMultiplayerFit(modes: string[]): number {
  if (modes.length === 0) return 25;

  const normalized = modes.join(" ").toLowerCase();
  let score = 48;
  if (normalized.includes("co-op") || normalized.includes("cooperative")) {
    score += 18;
  }
  if (normalized.includes("online") || normalized.includes("lan")) {
    score += 12;
  }
  if (normalized.includes("local") || normalized.includes("split screen")) {
    score += 12;
  }
  if (normalized.includes("pvp")) {
    score += 8;
  }
  if (modes.length >= 2) {
    score += 5;
  }

  return clamp(score, 0, 100);
}

function freshnessScore(releaseDate: string | null | undefined, today: Date) {
  const days = daysSinceRelease(releaseDate, today);
  if (days === null) return 35;
  if (days >= 0 && days <= 30) return 100;
  if (days <= 90) return 75;
  if (days <= 365) return 45;
  return 25;
}

function lightweightDiscoveryValue(
  facts: GameRecommendationFacts,
  reviewQuality: number,
  freshness: number,
) {
  const positiveReviewPct = facts.positiveReviewPct ?? 0;
  const totalReviews = facts.totalReviews ?? 0;
  const currentPlayers = facts.currentPlayers ?? 0;
  const sleeperScore =
    (reviewQuality >= 78 && totalReviews < 500 ? 25 : 0) +
    (currentPlayers < 300 && positiveReviewPct >= 85 ? 20 : 0) +
    (facts.demoStatus === "demo_only" || facts.demoStatus === "released_with_demo"
      ? 15
      : 0) +
    (freshness >= 75 ? 10 : 0);

  const demoPotential =
    facts.demoStatus === "demo_only"
      ? 60
      : facts.demoStatus === "released_with_demo"
        ? 45
        : facts.demoStatus === "released"
          ? 15
          : 10;

  return clamp(
    0.45 * freshness + 0.4 * sleeperScore + 0.15 * demoPotential,
    0,
    100,
  );
}

function lightweightConfidence(facts: GameRecommendationFacts) {
  const reviewConfidence = 1 - Math.exp(-(facts.totalReviews ?? 0) / 120);
  const modeConfidence = Math.min(facts.multiplayerModes.length / 3, 1);
  const activityConfidence =
    typeof facts.currentPlayers === "number" ? 1 : 0.35;

  return clamp(
    0.6 * reviewConfidence + 0.25 * activityConfidence + 0.15 * modeConfidence,
    0,
    1,
  );
}

function daysSinceRelease(
  releaseDate: string | null | undefined,
  today: Date,
): number | null {
  if (!releaseDate) return null;
  const release = new Date(`${releaseDate}T00:00:00Z`);
  if (Number.isNaN(release.getTime())) return null;
  const todayUtc = Date.UTC(
    today.getUTCFullYear(),
    today.getUTCMonth(),
    today.getUTCDate(),
  );
  return Math.floor((todayUtc - release.getTime()) / 86_400_000);
}

function roundOne(value: number): number {
  return Math.round(value * 10) / 10;
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

File: crates/mpgs-core/src/recommendation.rs
use serde::{Deserialize, Serialize};
use time::{format_description::FormatItem, macros::format_description, Date};

const ISO_DATE: &[FormatItem<'_>] = format_description!("[year]-[month]-[day]");

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DemoStatus {
    DemoOnly,
    ReleasedWithDemo,
    Released,
    Unknown,
}

impl DemoStatus {
    pub fn from_parts(is_demo_app: bool, has_demo: bool) -> Self {
        match (is_demo_app, has_demo) {
            (true, _) => Self::DemoOnly,
            (false, true) => Self::ReleasedWithDemo,
            (false, false) => Self::Released,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ReleaseBucket {
    New,
    Classic,
    ClassicHidden,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameFacts {
    pub appid: u32,
    pub name: String,
    pub release_date: Option<String>,
    pub positive_review_pct: Option<f64>,
    pub total_reviews: Option<u32>,
    pub current_players: Option<u32>,
    pub multiplayer_modes: Vec<String>,
    pub demo_status: DemoStatus,
    pub ai_score: Option<f64>,
}

pub fn compute_recommendation_score(facts: &GameFacts, today_iso: &str) -> f64 {
    if let Some(ai_score) = facts.ai_score {
        return round_one(ai_score.clamp(0.0, 100.0));
    }

    let review_quality = lightweight_review_quality(facts);
    let multiplayer_fit = lightweight_multiplayer_fit(&facts.multiplayer_modes);
    let freshness = freshness_score(facts.release_date.as_deref(), today_iso);
    let discovery_value = lightweight_discovery_value(facts, review_quality, freshness);
    let confidence = lightweight_confidence(facts);
    let uncertainty_penalty = (1.0 - confidence) * 10.0;
    let preanalysis_penalty = 4.0;

    let lightweight_quality_proxy =
        0.45 * review_quality + 0.30 * multiplayer_fit + 0.15 * freshness + 0.10 * discovery_value;

    round_one(
        (0.55 * lightweight_quality_proxy
            + 0.20 * multiplayer_fit
            + 0.15 * discovery_value
            + 0.10 * freshness
            - uncertainty_penalty
            - preanalysis_penalty)
            .clamp(0.0, 100.0),
    )
}

pub fn bucket_game(facts: &GameFacts, today_iso: &str) -> ReleaseBucket {
    match days_since_release(facts.release_date.as_deref(), today_iso) {
        Some(days) if (0..=30).contains(&days) => ReleaseBucket::New,
        Some(days) if days > 30 => {
            let review_pct = facts.positive_review_pct.unwrap_or_default();
            let total_reviews = facts.total_reviews.unwrap_or_default();
            if review_pct >= 80.0 && total_reviews >= 1_000 {
                ReleaseBucket::Classic
            } else if review_pct >= 60.0 && total_reviews >= 300 {
                ReleaseBucket::ClassicHidden
            } else {
                ReleaseBucket::ClassicHidden
            }
        }
        _ => ReleaseBucket::ClassicHidden,
    }
}

pub fn today_iso_utc() -> String {
    let now = std::time::SystemTime::now();
    let duration = now
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    let days = duration / 86_400;
    let date = Date::from_julian_day(2_440_588 + days as i32).unwrap_or(Date::MIN);
    date.format(ISO_DATE)
        .unwrap_or_else(|_| "2026-04-26".to_string())
}

fn lightweight_review_quality(facts: &GameFacts) -> f64 {
    let raw_positive_rate = facts.positive_review_pct.unwrap_or(0.0).clamp(0.0, 100.0) / 100.0;
    let total_reviews = facts.total_reviews.unwrap_or(0) as f64;
    let positive = total_reviews * raw_positive_rate;
    let bayes_positive_rate = (positive + 35.0) / (total_reviews + 35.0 + 15.0);
    let confidence = 1.0 - (-(total_reviews) / 120.0).exp();

    (100.0 * (bayes_positive_rate * 0.85 + raw_positive_rate * 0.15) * confidence
        + 55.0 * (1.0 - confidence))
        .clamp(0.0, 100.0)
}

fn lightweight_multiplayer_fit(modes: &[String]) -> f64 {
    if modes.is_empty() {
        return 25.0;
    }

    let normalized = modes
        .iter()
        .map(|mode| mode.to_ascii_lowercase())
        .collect::<Vec<_>>()
        .join(" ");

    let mut score: f64 = 48.0;
    if normalized.contains("co-op") || normalized.contains("cooperative") {
        score += 18.0;
    }
    if normalized.contains("online") || normalized.contains("lan") {
        score += 12.0;
    }
    if normalized.contains("local") || normalized.contains("split screen") {
        score += 12.0;
    }
    if normalized.contains("pvp") {
        score += 8.0;
    }
    if modes.len() >= 2 {
        score += 5.0;
    }

    score.clamp(0.0, 100.0)
}

fn freshness_score(release_date: Option<&str>, today_iso: &str) -> f64 {
    match days_since_release(release_date, today_iso) {
        Some(days) if (0..=30).contains(&days) => 100.0,
        Some(days) if (31..=90).contains(&days) => 75.0,
        Some(days) if (91..=365).contains(&days) => 45.0,
        Some(_) => 25.0,
        None => 35.0,
    }
}

fn lightweight_discovery_value(facts: &GameFacts, review_quality: f64, freshness: f64) -> f64 {
    let positive_review_pct = facts.positive_review_pct.unwrap_or(0.0);
    let total_reviews = facts.total_reviews.unwrap_or(0);
    let current_players = facts.current_players.unwrap_or(0);
    let sleeper_score = if review_quality >= 78.0 && total_reviews < 500 {
        25.0
    } else {
        0.0
    } + if current_players < 300 && positive_review_pct >= 85.0 {
        20.0
    } else {
        0.0
    } + if matches!(
        facts.demo_status,
        DemoStatus::DemoOnly | DemoStatus::ReleasedWithDemo
    ) {
        15.0
    } else {
        0.0
    } + if freshness >= 75.0 { 10.0 } else { 0.0 };
    let demo_potential = match facts.demo_status {
        DemoStatus::DemoOnly => 60.0,
        DemoStatus::ReleasedWithDemo => 45.0,
        DemoStatus::Released => 15.0,
        DemoStatus::Unknown => 10.0,
    };

    (0.45 * freshness + 0.40 * sleeper_score + 0.15 * demo_potential).clamp(0.0, 100.0)
}

fn lightweight_confidence(facts: &GameFacts) -> f64 {
    let review_confidence = 1.0 - (-(facts.total_reviews.unwrap_or(0) as f64) / 120.0).exp();
    let mode_confidence = (facts.multiplayer_modes.len() as f64 / 3.0).min(1.0);
    let activity_confidence = if facts.current_players.is_some() {
        1.0
    } else {
        0.35
    };

    (0.60 * review_confidence + 0.25 * activity_confidence + 0.15 * mode_confidence).clamp(0.0, 1.0)
}

fn days_since_release(release_date: Option<&str>, today_iso: &str) -> Option<i64> {
    let release = Date::parse(release_date?, ISO_DATE).ok()?;
    let today = Date::parse(today_iso, ISO_DATE).ok()?;
    Some((today - release).whole_days())
}

fn round_one(value: f64) -> f64 {
    (value * 10.0).round() / 10.0
}
